kvserver: deflake, fix, speed up TestTenantRateLimiter

tbg opened the pull request:
This test would occasionally (rarely) hang for 120s and then fail with
an opaque AmbiguousResultError in case that some of the requests the
test doesn't expect to block end up doing so.

I'm fairly sure this test can sometimes deadlock because enough extra
writes sneak in to cause the test's requests to block.

Seeing how this is a tarpit test (once you start looking your work day
is consumed), I addressed the following problems:

- it claimed to verify metrics, but the regexp had rotted
  and so it wasn't verifying anything.
- it ran into throttling in the client-side tenant rate limiter,
  so the test took 10s-15s. Now it runs in ~1s.
- it also did >3k requests, so it was quite expensive. Now
  it does a couple hundred bigger ones, which should be a
  smidge cheaper.
- it was flaky due to trying to make very strong claims about
  when requests would block, despite dozens of background requests
  that are hard to avoid sneaking in on every run.
- it would be difficult to debug when hanging due to unexpected
  blocking.

Now it's a somewhat weaker but still meaningful test that going forward
we are paying a much smaller tax for. Should it block again the test
will fail gracefully; it will still be difficult to root cause but
it is also now wholly unexpected that this should ever occur.

See: https://github.com/cockroachdb/cockroach/issues/129612#issuecomment-2310365016

Fixes #129612.

Release note: None
Epic: None

tbg committed:
kvserver: deflake, fix, speed up TestTenantRateLimiter

This test would occasionally (rarely) hang for 120s and then fail with
an opaque AmbiguousResultError in case that some of the requests the
test doesn't expect to block end up doing so.

I'm fairly sure this test can sometimes deadlock because enough extra
writes sneak in to cause the test's requests to block.

Seeing how this is a tarpit test (once you start looking your work day
is consumed), I addressed the following problems:

- it claimed to verify metrics, but the regexp had rotted
  and so it wasn't verifying anything.
- it ran into throttling in the client-side tenant rate limiter,
  so the test took 10s-15s. Now it runs in ~1s.
- it also did >3k requests, so it was quite expensive. Now
  it does a couple hundred bigger ones, which should be a
  smidge cheaper.
- it was flaky due to trying to make very strong claims about
  when requests would block, despite dozens of background requests
  that are hard to avoid sneaking in on every run.
- it would be difficult to debug when hanging due to unexpected
  blocking.

Now it's a somewhat weaker but still meaningful test that going forward
we are paying a much smaller tax for. Should it block again the test
will fail gracefully; it will still be difficult to root cause but
it is also now wholly unexpected that this should ever occur.

See: https://github.com/cockroachdb/cockroach/issues/129612#issuecomment-2310365016

Fixes #129612.

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/129640)
<!-- Reviewable:end -->

tbg committed:
tenantrate: generalize quotapool testing knob

This will be used in an upcoming change to TestTenantRateLimiter
to inject additional options into the tenantrate quota pool.

Release note: None
Epic: None

arulajmani reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

Thanks for improving this test. 

> Reviewed 5 of 5 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/129640#-:-O66boYe9Sjft__8gu40:bdk9rb0)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @tbg)



<!-- Sent from Reviewable.io -->

tbg commented:
bors r+

TFTR!

arulajmani commented:
blathers backport 23.2 24.1 24.2
